MAGPI PlatformIO/src/main.cpp

```diff
@@ -0,0 +1,461 @@
+
+  } else{
+    current_time = since_last_GPS - timestamp_GPS_package; //////////need check//////////// should use gps time + micros OR just treat it as another variable
+  }
+
+  mpu.getEvent(&a, &w, &temp);
+  bmp.performReading(); //write to public attributes temperature and pressure
+  alt_bmp = bmp.readAltitude(SEALEVELPRESSURE_HPA); 
+  #endif
+
+  #ifdef DEBUG_SERIAL_INPUT
+    serial_debug("lat: ");
+    lat = double(Serial.parseFloat());
+    serial_debug("lon: ");
+    lon = double(Serial.parseFloat());
+    serial_debug("GPS Time (HHMMSSCC): ");
+    current_time = Serial.parseInt(); //UNIX time
+    //gps_alt = myGPS.altitude.meters();
+    serial_debug("heading: ");
+    heading = myGPS.course.deg();
+
+    //w.gyro.x = 0.0;
+
+    //speed = myGPS.speed.mps();
+  #endif
+}
+
+void save_data(){
+  //Use Serial for DEBUG [add pause function]
+  myFile = SD.open(filename, FILE_WRITE);
+  myFile.print(current_time);
+  myFile.print(delim);
+
+  //mpu data sensors readings
+  myFile.print(a.acceleration.x); myFile.print(delim);
+  myFile.print(a.acceleration.y); myFile.print(delim);
+  myFile.print(a.acceleration.z); myFile.print(delim);
+  myFile.print(w.gyro.x); myFile.print(delim);
+  myFile.print(w.gyro.y); myFile.print(delim);
+  myFile.print(w.gyro.z); myFile.print(delim);
+
+  //bmp
+  myFile.print(bmp.temperature); myFile.print(delim);
+  myFile.print(bmp.pressure); myFile.print(delim);
+  myFile.print(alt_bmp); myFile.print(delim);
+
+  myFile.print(lat); myFile.print(delim);
+  myFile.print(lon); myFile.println(""); 
+
+  myFile.close(); 
+  
+}
+
+double principal_angle(double angle){
+    if (angle > 180.0){
+      angle += -360.0;
+    } else if (angle < -180.0){
+      angle += 360.0;
+    } else if (angle == 360.0 || angle == -360.0){
+      angle = 0.0;
+    }
+  return angle;
+}
+
+void rotate_motor(double length){
+  int pos = round(turn_per_length * length);
+  digitalWrite(EN, HIGH);
+  
+  if (pos > 0){
+
+    digitalWrite(A1_3, HIGH);
+    while(motor.read() < pos);
+
+  } else if (pos < 0){ // e.g. pos_now = 0, set point < -10 => pos_now > set point then turn left
+    
+    digitalWrite(A2_4, HIGH);
+    while(motor.read() > pos);
+
+  }
+  digitalWrite(EN, LOW);
+  digitalWrite(A1_3, LOW);
+  digitalWrite(A2_4, LOW);
+  
+}
+
+void algorithm(){
+
+  //int turn_no = 0;
+  int turn_reset = 0;
+  double yaw_req = 0.0;
+  target_heading = myGPS.courseTo(lat, lon, TARGET_LAT, TARGET_LON);
+  yaw_req = (target_heading - heading);
+  yaw_req = principal_angle(yaw_req);
+  
+// get yaw difference -> P controller
+// P controller needs -> current yaw               /////need to set the map of the pulley////////// 
+
+  yaw = 0.0;
+  elapsedMicros algorithm_time; //////////change back to elasped micros   //reset algorithm_time to find time passed and for integration
+
+  while (!(((error = yaw_req - yaw)<= yaw_control_cutoff)&&(error >= -yaw_control_cutoff)) && (algorithm_time < control_timeout)) {
```

Comment:
I think this is equivalent right? All you're checking is if the error is small enough to stop.

```suggestion
  while (!(abs(error = yaw_req - yaw) <= yaw_control_cutoff)) && (algorithm_time < control_timeout)) {
```

Reply:
I will change the code to this, totally forgot abs() exist